app/NodeIPC/Main.hs

```diff
@@ -1,12 +1,57 @@
+{-# LANGUAGE OverloadedStrings #-}
+
 module Main
     ( main
     ) where
 
 import           Cardano.Prelude
-import           Cardano.Shell.NodeIPC (Port (..), ProtocolDuration (..),
-                                        startNodeJsIPC)
+
+import           Cardano.Shell.NodeIPC (MsgIn (..),
+                                        Port (..), ProtocolDuration (..),
+                                        ReadHandle (..), WriteHandle (..),
+                                        sendMessage, startIPC, startNodeJsIPC)
+import           GHC.IO.Handle.FD (fdToHandle)
+import           Prelude (String, error)
+import           System.Environment
+import           System.IO (BufferMode (..), hClose, hSetBuffering)
+import           System.Process (createPipe)
+
+port :: Port
+port = Port 8090
 
 main :: IO ()
 main = do
-    let port = Port 8090
-    startNodeJsIPC SingleMessage port
+    (cmd:_) <- getArgs
+    case cmd of
+        "js"      -> startNodeJsIPC SingleMessage port
+        "haskell" -> bracket acquire restore action
+        _         -> return ()
+  where
+    acquire :: IO (Handle, Handle, Handle)
+    acquire = do
+        clientWHandle <- getHandleFromEnv "FD_WRITE_HANDLE"
+        hSetBuffering clientWHandle LineBuffering
+        (serverRHandle , serverWHandle) <- createPipe
+        return (clientWHandle, serverRHandle, serverWHandle)
+
+    restore :: (Handle, Handle, Handle) -> IO ()
+    restore (clientWHandle, serverRHandle, serverWHandle) = do
+        hClose clientWHandle
+        hClose serverRHandle
+        hClose serverWHandle
+
+    action :: (Handle, Handle, Handle) -> IO ()
+    action (clientWHandle, serverRHandle, serverWHandle) = do
+        let clientWriteHandle = WriteHandle clientWHandle
+        let serverReadHandle  = ReadHandle serverRHandle
+        sendMessage (WriteHandle serverWHandle) Ping
+        startIPC SingleMessage serverReadHandle clientWriteHandle port
+
+    getHandleFromEnv :: String -> IO Handle
+    getHandleFromEnv envName = do
+        mFdstring <- lookupEnv envName
+        case mFdstring of
+            Nothing -> error $ "Unable to find fd: " <> envName
+            Just fdstring -> case readEither fdstring of
+                Left err -> error $ "Could not parse file descriptor: " <> toS err
```

Comment:
I am not a big fan of error handling via just throwing an `error`. It works acceptable well for small projects and demos, but scales very poorly to larger projects where you can suddenly have 100s of calls to `error`.

I think it would be much better to have the type of `getHandleFromEnv` to be:
```
    getHandleFromEnv :: String -> IO (Either NodeIPCError Handle)
```
and then have `NodeIPCError` be a sum type with a proper render function (`NodeIPCError -> Text`) that can be used to print the error.

Reply:
"The illusion of certainty" - using either the `error` or the `Either` here will result in the application "crashing", right?
In either case, it's a non-recoverable exception, right - does `Either` bring something here other than the delay of the exception - the handler of the `Either` will have to do something like `error` or `panic`?
It's nice for testing, having "pure" exceptions, reasoning about the function via its type and it might be easier to reason about in some cases, but sometimes the result is the same - when you have a `Left`, you shutdown with an error message.
While I agree in principle, I always wonder what/where is the difference, since the end result is the same in cases like these? It does make things complicated when you have it passing that signature around and at some point, you do the same thing you could have done in the first place.
:hear_no_evil:

Reply:
@ksaric I would agree for how this is structured at the moment, as a standalone application.

However, my experience suggests that anything other than a demo should be a library and a minimal wrapper to turn it into an application. Ideally, this thin wrapper should contain nothing more than command line parsing and calls into the library.

I hope its rather obvious that library code should never call `error` (or more controversially, throw an exception), but rather pass errors back to the application, preferably as the `Left` of an `Either`. 
The library of course should not expose functions that use `ExceptT`, but certainly can use `ExceptT` internally.

The reasons to make libraries:
* Code designed as a library is usually better structured than application code and will usually have a real API.
* Code build as a library for one client program can be easily used by another client, probably without rework.
* Code built as a library is easier to test than code built as an application.

The reasons to use `ExceptT` rather than exceptions:
* With `ExceptT` the errors a function can have are explicitly listed in the type signature.
* Errors types (used with `Either` and `ExceptT`) are closed sets whereas the exception system was designed to be (and is) extensible.
* Errors types (used with `Either` and `ExceptT`) compose (data types within data types within data types ....) and the outer error data type is still a closed set. The exceptions system allows composition via its extensibility, but the open extensibility means there is no way for the compiler to ensure that all exceptions are handled.

Reply:
That is a very good point indeed. 
Visibility/Explicitness is important, especially when you want to use something that you didn't write yourself and don't know the execution paths.